prov/efa/src/efa_cq.c

```diff
@@ -331,11 +340,119 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
 	ofi_bufpool_destroy(cq->wce_pool);
 err_destroy_cq:
-	ibv_destroy_cq(cq->ibv_cq);
+	ibv_destroy_cq(ibv_cq_ex_to_cq(cq->ibv_cq_ex));
 err_free_util_cq:
 	ofi_cq_cleanup(&cq->util_cq);
 err_free_cq:
 	free(cq);
 	return ret;
 }
 
+
+#if EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+
+struct readerr_input {
+	struct fi_cq_err_entry *entry;
+	struct efa_cq *efa_cq;
+};
+
+static enum ibv_wc_opcode _read_send_code(struct ibv_cq_ex *current)
+{
+	return IBV_WC_SEND;
+};
+
+static int setup_readerr(void **state)
+{
+	static struct fid_cq fid_cq = {0};
+	static struct util_cq util_cq = {0};
+	static struct efa_domain domain = {0};
+	static struct efa_fabric fabric = {0};
+	static struct util_fabric util_fabric = {0};
+	static struct fid_fabric fid_fabric = {0};
+	static struct ibv_cq_ex *ibv_cq_ex;
+	static struct efa_cq *efa_cq;
+	static struct fi_cq_err_entry *fi_cq_err_entry;
+	static struct readerr_input *readerr_input;
+
+	ibv_cq_ex = calloc(1, sizeof(struct ibv_cq_ex));
+	efa_cq = calloc(1, sizeof(struct efa_cq));
+	fi_cq_err_entry = calloc(1, sizeof(struct fi_cq_err_entry));
+	readerr_input = calloc(1, sizeof(struct readerr_input));
+
+	if (!ibv_cq_ex || !efa_cq || !fi_cq_err_entry || !readerr_input) {
+		return -1;
+	}
+
+	fid_fabric.api_version = 0;
+	util_fabric.fabric_fid = fid_fabric;
+	fabric.util_fabric = util_fabric;
+	domain.fabric = &fabric;
+	ibv_cq_ex->read_opcode = &_read_send_code;
+	efa_cq->util_cq = util_cq;
+	efa_cq->util_cq.cq_fid = fid_cq;
+	efa_cq->ibv_cq_ex = ibv_cq_ex;
+	efa_cq->domain = &domain;
+	ofi_spin_init(&efa_cq->lock);
+
+	readerr_input->entry = fi_cq_err_entry;
+	readerr_input->efa_cq = efa_cq;
+
+	*state = readerr_input;
+
+	return 0;
+}
+
+static int teardown_readerr(void **state)
+{
+	struct readerr_input *input = *state;
+
+	ofi_spin_destroy(&input->efa_cq->lock);
+	free(input->entry);
+	free(input->efa_cq->ibv_cq_ex);
+	free(*state);
+
+	return 0;
+}
+
+static void test_efa_cq_readerr_happy(void **state) {
+	ssize_t rc;
+	struct readerr_input *input = *state;
+
+	input->efa_cq->ibv_cq_ex->wr_id = 0;
+	input->efa_cq->ibv_cq_ex->status = 1;
+
+	rc = efa_cq_readerr(&input->efa_cq->util_cq.cq_fid, input->entry, 0);
+
+	assert_return_code(rc, 1);
+}
+
+static void test_efa_cq_readerr_sad(void **state) {
+	ssize_t rc;
+	struct readerr_input *input = *state;
+
+	input->efa_cq->ibv_cq_ex->wr_id = 0;
+	input->efa_cq->ibv_cq_ex->status = 0; /* error should have non-zero status */
+
+	rc = efa_cq_readerr(&input->efa_cq->util_cq.cq_fid, input->entry, 0);
+
+	assert_int_equal(rc, -FI_EAGAIN);
+}
+
+int run_efa_cq_tests(void) {
+	int ret;
+
+	cmocka_set_message_output(CM_OUTPUT_XML);
+
+	const struct CMUnitTest efa_cq_unit_tests[] = {
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_happy, setup_readerr, teardown_readerr),
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_sad, setup_readerr, teardown_readerr),
+	};
+
+	ret = cmocka_run_group_tests_name("efa cq unit tests", efa_cq_unit_tests, NULL, NULL);
```

Comment:
why create a new group instead of add the test to existing group?

Reply:
I don't see a strong connection between these tests, and they are testing distinct aspects from what I can tell.

The more important reason is that, I find the recommended approach not scalable as we add more tests, i.e.
```
1. Add unit tests to the bottom of the c source file you wish to test.
2. Add headers to efa_unit_tests.h for each new unit test.
3. Create or find an existing cmocka test group in efa_unit_tests.c
4. Add your new unit test to the cmocka test group.
```

For functions called in `efa_unit_tests.c` we will need to declare them in `efa_unit_tests.h`, which requires more code. This is also restrictive for developers, e.g. it's hard to reuse static functions in `efa_xxx.c` (declaring static functions in `efa_unit_tests.h` is not as straight forward because of dependencies).

From my perspective it's better to be frugal with what we put in `efa_unit_tests.x`, and hide detailed test fixtures in respective source files. Modifying tests is easier when the related code is localized in one or two places and shared with few other tests.

Reply:
> I don't see a strong connection between these tests, and they are testing distinct aspects from what I can tell.

They are tests of EFA provider, therefore is in a group named "efa unit tests". It would be confusing to see two groups of tests, one is named "efa unit tests", the other is named "efa cq unit tests". 

> For functions called in efa_unit_tests.c we will need to declare them in efa_unit_tests.h, which requires more code. 

Yes, that is 1 line for each test, and one will need to declare a function so that it can be called from `efa_unit_tests.c`

IMO this PR added more code (and confusion) to avoid that declaration. The function `run_efa_cq_tests` (15 lines of code) can be avoided if we add the test directly to existing group.

Reply:
The existing `efa unit tests` should be split into smaller groups. Even though they all 'test efa provider', groups are logically independent from each other. For example, the [cmocka_run_group_tests()](https://api.cmocka.org/group__cmocka__exec.html#ga7c62fd0acf2235ce98268c28ee262a57) function has this signature:
```c
int cmocka_run_group_tests ( 
                const struct CMUnitTest group_tests[],
		CMFixtureFunction  	group_setup,
		CMFixtureFunction  	group_teardown 
)
```
It implies that one criteria for test group is that, tests in the same group should require the same setup and teardown process. Tests in this PR do not align with the existing tests and has its own requirements (e.g. the new tests do not require EFA device).

More importantly keeping all test related code in `efa_unit_tests` is not a scalable approach and leads to unnecessary code - yes, centralized code is easier to reuse, but enforcing it means the developer must share details though that's not used elsewhere. Instead the developer should decide the right level of encapsulation, and what she wants to expose.

Reply:
> The existing efa unit tests should be split into smaller groups.

There are only 2 test in `efa unit tests`. I think split them into smaller groups is overkill.

> More importantly keeping all test related code in efa_unit_tests is not a scalable approach and leads to unnecessary code.

I am not sure I agree. In my opinion the current PR has unnecessary code: the function `run_efa_cq_test` is not necessary.